File: src/robot_wire_protocol.c
#include "robot_wire_protocol.h"

static void write_u16_le(uint8_t* output, uint16_t value)
{
    output[0] = (uint8_t)(value & UINT16_C(0x00ff));
    output[1] = (uint8_t)((value >> 8U) & UINT16_C(0x00ff));
}

static void write_u32_le(uint8_t* output, uint32_t value)
{
    output[0] = (uint8_t)(value & UINT32_C(0x000000ff));
    output[1] = (uint8_t)((value >> 8U) & UINT32_C(0x000000ff));
    output[2] = (uint8_t)((value >> 16U) & UINT32_C(0x000000ff));
    output[3] = (uint8_t)((value >> 24U) & UINT32_C(0x000000ff));
}

static void write_u64_le(uint8_t* output, uint64_t value)
{
    size_t index;
    for (index = 0U; index < 8U; ++index) {
        output[index] = (uint8_t)((value >> (8U * index)) & UINT64_C(0xff));
    }
}

static uint16_t read_u16_le(const uint8_t* input)
{
    return (uint16_t)((uint16_t)input[0] | ((uint16_t)input[1] << 8U));
}

static uint32_t read_u32_le(const uint8_t* input)
{
    return (uint32_t)((uint32_t)input[0] | ((uint32_t)input[1] << 8U) |
                      ((uint32_t)input[2] << 16U) | ((uint32_t)input[3] << 24U));
}

static uint64_t read_u64_le(const uint8_t* input)
{
    uint64_t value = UINT64_C(0);
    size_t index;
    for (index = 0U; index < 8U; ++index) {
        value |= (uint64_t)input[index] << (8U * index);
    }
    return value;
}

static void encode_header(uint8_t* output, uint16_t message_type, uint16_t payload_size,
                          uint32_t sequence)
{
    write_u32_le(output, ROBOT_WIRE_MAGIC);
    write_u16_le(&output[4], ROBOT_WIRE_PROTOCOL_VERSION);
    write_u16_le(&output[6], message_type);
    write_u16_le(&output[8], payload_size);
    write_u16_le(&output[10], UINT16_C(0));
    write_u32_le(&output[12], sequence);
}

static robot_wire_result_t decode_header(const uint8_t* input, size_t input_size,
                                          uint16_t expected_message_type,
                                          uint16_t expected_payload_size, uint32_t* sequence)
{
    if (input == NULL || sequence == NULL) return ROBOT_WIRE_NULL_ARGUMENT;
    if (input_size != ROBOT_WIRE_HEADER_SIZE + (size_t)expected_payload_size) {
        return ROBOT_WIRE_INVALID_LENGTH;
    }
    if (read_u32_le(input) != ROBOT_WIRE_MAGIC) return ROBOT_WIRE_INVALID_MAGIC;
    if (read_u16_le(&input[4]) != ROBOT_WIRE_PROTOCOL_VERSION) return ROBOT_WIRE_INVALID_VERSION;
    if (read_u16_le(&input[6]) != expected_message_type) return ROBOT_WIRE_INVALID_MESSAGE_TYPE;
    if (read_u16_le(&input[8]) != expected_payload_size) return ROBOT_WIRE_INVALID_LENGTH;
    *sequence = read_u32_le(&input[12]);
    return ROBOT_WIRE_OK;
}

static void encode_rgb_values(const robot_rgb_t* lights, uint8_t* output)
{
    size_t index;
    for (index = 0U; index < ROBOT_LIGHT_COUNT; ++index) {
        const size_t offset = index * 3U;
        output[offset] = lights[index].red;
        output[offset + 1U] = lights[index].green;
        output[offset + 2U] = lights[index].blue;
    }
}

static void decode_rgb_values(const uint8_t* input, robot_rgb_t* lights)
{
    size_t index;
    for (index = 0U; index < ROBOT_LIGHT_COUNT; ++index) {
        const size_t offset = index * 3U;
        lights[index].red = input[offset];
        lights[index].green = input[offset + 1U];
        lights[index].blue = input[offset + 2U];
    }
}

robot_wire_result_t robot_wire_encode_lights_command(const robot_lights_command_t* command,
                                                      uint8_t* output, size_t output_size)
{
    if (command == NULL || output == NULL) return ROBOT_WIRE_NULL_ARGUMENT;
    if (output_size < ROBOT_LIGHTS_MESSAGE_SIZE) return ROBOT_WIRE_BUFFER_TOO_SMALL;
    encode_header(output, ROBOT_WIRE_MESSAGE_LIGHTS_COMMAND, (uint16_t)ROBOT_LIGHTS_PAYLOAD_SIZE,
                  command->sequence);
    output[16] = (uint8_t)ROBOT_LIGHT_COUNT;
    output[17] = UINT8_C(0);
    output[18] = UINT8_C(0);
    output[19] = UINT8_C(0);
    encode_rgb_values(command->lights, &output[20]);
    return ROBOT_WIRE_OK;
}

robot_wire_result_t robot_wire_decode_lights_command(const uint8_t* input, size_t input_size,
                                                      robot_lights_command_t* command)
{
    robot_wire_result_t result;
    uint32_t sequence;
    if (command == NULL) return ROBOT_WIRE_NULL_ARGUMENT;
    result = decode_header(input, input_size, ROBOT_WIRE_MESSAGE_LIGHTS_COMMAND,
                           (uint16_t)ROBOT_LIGHTS_PAYLOAD_SIZE, &sequence);
    if (result != ROBOT_WIRE_OK) return result;
    if (input[16] != (uint8_t)ROBOT_LIGHT_COUNT) return ROBOT_WIRE_INVALID_LIGHT_COUNT;
    command->sequence = sequence;
    decode_rgb_values(&input[20], command->lights);
    return ROBOT_WIRE_OK;
}

robot_wire_result_t robot_wire_encode_lights_state(const robot_lights_state_t* state,
                                                    uint8_t* output, size_t output_size)
{
    if (state == NULL || output == NULL) return ROBOT_WIRE_NULL_ARGUMENT;
    if (output_size < ROBOT_LIGHTS_MESSAGE_SIZE) return ROBOT_WIRE_BUFFER_TOO_SMALL;
    if (state->result > (uint8_t)ROBOT_LIGHTS_RESULT_BUSY_SOURCE) return ROBOT_WIRE_INVALID_RESULT;
    if (state->active_source > (uint8_t)ROBOT_LIGHTS_SOURCE_EDUCATION_JSON) return ROBOT_WIRE_INVALID_SOURCE;
    encode_header(output, ROBOT_WIRE_MESSAGE_LIGHTS_STATE, (uint16_t)ROBOT_LIGHTS_PAYLOAD_SIZE,
                  state->sequence);
    output[16] = (uint8_t)ROBOT_LIGHT_COUNT;
    output[17] = state->result;
    output[18] = state->active_source;
    output[19] = UINT8_C(0);
    encode_rgb_values(state->lights, &output[20]);
    return ROBOT_WIRE_OK;
}

robot_wire_result_t robot_wire_decode_lights_state(const uint8_t* input, size_t input_size,
                                                    robot_lights_state_t* state)
{
    robot_wire_result_t result;
    uint32_t sequence;
    if (state == NULL) return ROBOT_WIRE_NULL_ARGUMENT;
    result = decode_header(input, input_size, ROBOT_WIRE_MESSAGE_LIGHTS_STATE,
                           (uint16_t)ROBOT_LIGHTS_PAYLOAD_SIZE, &sequence);
    if (result != ROBOT_WIRE_OK) return result;
    if (input[16] != (uint8_t)ROBOT_LIGHT_COUNT) return ROBOT_WIRE_INVALID_LIGHT_COUNT;
    if (input[17] > (uint8_t)ROBOT_LIGHTS_RESULT_BUSY_SOURCE) return ROBOT_WIRE_INVALID_RESULT;
    if (input[18] > (uint8_t)ROBOT_LIGHTS_SOURCE_EDUCATION_JSON) return ROBOT_WIRE_INVALID_SOURCE;
    state->sequence = sequence;
    state->result = input[17];
    state->active_source = input[18];
    decode_rgb_values(&input[20], state->lights);
    return ROBOT_WIRE_OK;
}

/* A level or changed bit outside the valid mask would claim a signal the
 * vehicle does not implement, which is exactly the confusion the valid mask
 * exists to prevent. Rejecting rather than silently masking keeps encode and
 * decode inverses of each other, and matches how this codec treats every other
 * out-of-range field. */
static robot_wire_result_t check_pinout_masks(uint64_t level, uint64_t changed, uint64_t valid)
{
    if ((level & ~valid) != UINT64_C(0)) return ROBOT_WIRE_INVALID_PINOUT_MASK;
    if ((changed & ~valid) != UINT64_C(0)) return ROBOT_WIRE_INVALID_PINOUT_MASK;
    return ROBOT_WIRE_OK;
}

robot_wire_result_t robot_wire_encode_pinout_event(const robot_pinout_event_t* event,
                                                    uint8_t* output, size_t output_size)
{
    robot_wire_result_t result;
    if (event == NULL || output == NULL) return ROBOT_WIRE_NULL_ARGUMENT;
    if (output_size < ROBOT_PINOUT_EVENT_MESSAGE_SIZE) return ROBOT_WIRE_BUFFER_TOO_SMALL;
    result = check_pinout_masks(event->inputs_level, event->inputs_changed, event->inputs_valid);
    if (result != ROBOT_WIRE_OK) return result;
    result = check_pinout_masks(event->outputs_level, event->outputs_changed, event->outputs_valid);
    if (result != ROBOT_WIRE_OK) return result;
    /* Header sequence is zero: an event answers no request. */
    encode_header(output, ROBOT_WIRE_MESSAGE_PINOUT_EVENT,
                  (uint16_t)ROBOT_PINOUT_EVENT_PAYLOAD_SIZE, UINT32_C(0));
    write_u64_le(&output[16], event->uptime_ms);
    write_u32_le(&output[24], event->event_sequence);
    write_u32_le(&output[28], event->boot_id);
    write_u64_le(&output[32], event->inputs_level);
    write_u64_le(&output[40], event->inputs_changed);
    write_u64_le(&output[48], event->inputs_valid);
    write_u64_le(&output[56], event->outputs_level);
    write_u64_le(&output[64], event->outputs_changed);
    write_u64_le(&output[72], event->outputs_valid);
    return ROBOT_WIRE_OK;
}

robot_wire_result_t robot_wire_decode_pinout_event(const uint8_t* input, size_t input_size,
                                                    robot_pinout_event_t* event)
{
    robot_wire_result_t result;
    uint32_t header_sequence;
    robot_pinout_event_t decoded;
    if (event == NULL) return ROBOT_WIRE_NULL_ARGUMENT;
    result = decode_header(input, input_size, ROBOT_WIRE_MESSAGE_PINOUT_EVENT,
                           (uint16_t)ROBOT_PINOUT_EVENT_PAYLOAD_SIZE, &header_sequence);
    if (result != ROBOT_WIRE_OK) return result;
    decoded.uptime_ms = read_u64_le(&input[16]);
    decoded.event_sequence = read_u32_le(&input[24]);
    decoded.boot_id = read_u32_le(&input[28]);
    decoded.inputs_level = read_u64_le(&input[32]);
    decoded.inputs_changed = read_u64_le(&input[40]);
    decoded.inputs_valid = read_u64_le(&input[48]);
    decoded.outputs_level = read_u64_le(&input[56]);
    decoded.outputs_changed = read_u64_le(&input[64]);
    decoded.outputs_valid = read_u64_le(&input[72]);
    result = check_pinout_masks(decoded.inputs_level, decoded.inputs_changed, decoded.inputs_valid);
    if (result != ROBOT_WIRE_OK) return result;
    result =
        check_pinout_masks(decoded.outputs_level, decoded.outputs_changed, decoded.outputs_valid);
    if (result != ROBOT_WIRE_OK) return result;
    /* Populated only once every field has passed, so a rejected frame never
     * leaves the caller's struct half written. */
    *event = decoded;
    return ROBOT_WIRE_OK;
}

robot_wire_result_t robot_wire_encode_estop_clear_reset_request(
    const robot_estop_clear_reset_request_t* request, uint8_t* output, size_t output_size)
{
    if (request == NULL || output == NULL) return ROBOT_WIRE_NULL_ARGUMENT;
    if (output_size < ROBOT_ESTOP_CLEAR_RESET_MESSAGE_SIZE) return ROBOT_WIRE_BUFFER_TOO_SMALL;
    encode_header(output, ROBOT_WIRE_MESSAGE_ESTOP_CLEAR_RESET_REQUEST,
                  (uint16_t)ROBOT_ESTOP_CLEAR_RESET_PAYLOAD_SIZE, request->sequence);
    write_u32_le(&output[16], UINT32_C(0));
    return ROBOT_WIRE_OK;
}

robot_wire_result_t robot_wire_decode_estop_clear_reset_request(
    const uint8_t* input, size_t input_size, robot_estop_clear_reset_request_t* request)
{
    if (request == NULL) return ROBOT_WIRE_NULL_ARGUMENT;
    return decode_header(input, input_size, ROBOT_WIRE_MESSAGE_ESTOP_CLEAR_RESET_REQUEST,
                         (uint16_t)ROBOT_ESTOP_CLEAR_RESET_PAYLOAD_SIZE, &request->sequence);
}

robot_wire_result_t robot_wire_encode_estop_clear_reset_state(
    const robot_estop_clear_reset_state_t* state, uint8_t* output, size_t output_size)
{
    if (state == NULL || output == NULL) return ROBOT_WIRE_NULL_ARGUMENT;
    if (output_size < ROBOT_ESTOP_CLEAR_RESET_MESSAGE_SIZE) return ROBOT_WIRE_BUFFER_TOO_SMALL;
    if (state->result > (uint8_t)ROBOT_ESTOP_CLEAR_RESET_INTERNAL_ERROR) {
        return ROBOT_WIRE_INVALID_ESTOP_CLEAR_RESET_RESULT;
    }
    encode_header(output, ROBOT_WIRE_MESSAGE_ESTOP_CLEAR_RESET_RESULT,
                  (uint16_t)ROBOT_ESTOP_CLEAR_RESET_PAYLOAD_SIZE, state->sequence);
    output[16] = state->result;
    output[17] = state->needs_reset;
    output[18] = state->estop_active;
    output[19] = UINT8_C(0);
    return ROBOT_WIRE_OK;
}

robot_wire_result_t robot_wire_decode_estop_clear_reset_state(
    const uint8_t* input, size_t input_size, robot_estop_clear_reset_state_t* state)
{
    robot_wire_result_t result;
    uint32_t sequence;
    if (state == NULL) return ROBOT_WIRE_NULL_ARGUMENT;
    result = decode_header(input, input_size, ROBOT_WIRE_MESSAGE_ESTOP_CLEAR_RESET_RESULT,
                           (uint16_t)ROBOT_ESTOP_CLEAR_RESET_PAYLOAD_SIZE, &sequence);
    if (result != ROBOT_WIRE_OK) return result;
    if (input[16] > (uint8_t)ROBOT_ESTOP_CLEAR_RESET_INTERNAL_ERROR) {
        return ROBOT_WIRE_INVALID_ESTOP_CLEAR_RESET_RESULT;
    }
    if (input[17] > UINT8_C(1) || input[18] > UINT8_C(1)) return ROBOT_WIRE_INVALID_LENGTH;
    state->sequence = sequence;
    state->result = input[16];
    state->needs_reset = input[17];
    state->estop_active = input[18];
    return ROBOT_WIRE_OK;
}

File: python/robot_wire_protocol.py
"""Versioned wire codecs, Python mirror of the C11 implementation in ``src/``.

This module exists so tooling and test scripts encode the same bytes as the
firmware and the gateway, without anyone hand-packing a struct in a consumer
repository. It is the same protocol expressed twice, in one repository, which
is the only arrangement where the two can be tested against each other.

Like the C code it is deliberately independent of ROS 2, Zenoh, ESP-IDF and
hardware drivers: it turns messages into bytes and back, nothing more. The
transport belongs to the caller.

``tests/test_python_matches_golden.py`` checks this module against the same
golden frame the C test asserts on, so a divergence between the two
implementations is a failing test rather than a mystery on the bench.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass, field
from enum import IntEnum
from typing import List

__all__ = [
    "MAGIC",
    "PROTOCOL_VERSION",
    "HEADER_SIZE",
    "LIGHT_COUNT",
    "MessageType",
    "WireError",
    "LightsResult",
    "LightsSource",
    "EstopClearResetResult",
    "PinoutInput",
    "PinoutOutput",
    "Rgb",
    "LightsCommand",
    "LightsState",
    "EstopClearResetRequest",
    "EstopClearResetState",
    "PinoutEvent",
    "encode_pinout_event",
    "decode_pinout_event",
    "encode_lights_command",
    "decode_lights_command",
    "encode_lights_state",
    "decode_lights_state",
    "encode_estop_clear_reset_request",
    "decode_estop_clear_reset_request",
    "encode_estop_clear_reset_state",
    "decode_estop_clear_reset_state",
]

MAGIC = 0x31434252  # "RBC1" little-endian
PROTOCOL_VERSION = 1
HEADER_SIZE = 16

LIGHT_COUNT = 4
LIGHTS_PAYLOAD_SIZE = 16
LIGHTS_MESSAGE_SIZE = HEADER_SIZE + LIGHTS_PAYLOAD_SIZE

ESTOP_CLEAR_RESET_PAYLOAD_SIZE = 4
ESTOP_CLEAR_RESET_MESSAGE_SIZE = HEADER_SIZE + ESTOP_CLEAR_RESET_PAYLOAD_SIZE

EVENT_PREFIX_SIZE = 16
PINOUT_EVENT_PAYLOAD_SIZE = 64
PINOUT_EVENT_MESSAGE_SIZE = HEADER_SIZE + PINOUT_EVENT_PAYLOAD_SIZE

_HEADER = struct.Struct("<IHHHHI")
_PINOUT_BODY = struct.Struct("<QIIQQQQQQ")

_UINT64_MAX = 0xFFFFFFFFFFFFFFFF


class MessageType(IntEnum):
    LIGHTS_COMMAND = 3
    LIGHTS_STATE = 4
    ESTOP_CLEAR_RESET_REQUEST = 5
    ESTOP_CLEAR_RESET_RESULT = 6
    PINOUT_EVENT = 7


class LightsResult(IntEnum):
    OK = 0
    INVALID_COMMAND = 1
    APPLY_FAILED = 2
    OVERRIDDEN = 3
    BUSY_SOURCE = 4


class LightsSource(IntEnum):
    NONE = 0
    SAFETY = 1
    ROS_BINARY = 2
    EDUCATION_JSON = 3


class EstopClearResetResult(IntEnum):
    OK = 0
    ESTOP_ACTIVE = 1
    NO_RESET_PENDING = 2
    SAFETY_SEQUENCE_INCOMPLETE = 3
    FAULT_ACTIVE = 4
    INTERNAL_ERROR = 5


class PinoutInput(IntEnum):
    """Input bit numbers. See ``robot_pinout_input_bit_t`` and PINOUT_PROTOCOL.md.

    A bit names a signal rather than an expander pin, and is 1 when the named
    condition holds regardless of how the line is wired electrically.
    """

    # 0..15 operator panel
    BUTTON_UP = 0
    BUTTON_DOWN = 1
    BUTTON_ESC = 2
    BUTTON_ENTER = 3
    BUTTON_TEST = 4
    TOGGLE_1 = 5
    TOGGLE_2 = 6
    TOGGLE_3 = 7
    # 16..31 safety chain
    ESTOP = 16


class PinoutOutput(IntEnum):
    """Output bit numbers. See ``robot_pinout_output_bit_t``."""

    # 0..15 signalling and indication
    HORN = 0
    HEARTBEAT = 1
    FULLSTOP_LED = 2
    ZENOH_LED = 3
    INDICATOR_LEFT = 4
    INDICATOR_RIGHT = 5
    AUTONOMOUS_LED = 6
    # 16..31 drivetrain control
    THROTTLE_DIR_REVERSE = 16
    EBRAKE = 17
    SKID_DIR_FL_REVERSE = 18
    SKID_DIR_FR_REVERSE = 19
    SKID_DIR_RL_REVERSE = 20
    SKID_DIR_RR_REVERSE = 21


class WireError(ValueError):
    """Raised when a frame cannot be decoded.

    The C API returns a result enum; Python raises, because a caller that
    ignores a return value here would act on an uninitialised message. The
    ``reason`` matches the C enumerator name so both sides report a rejection
    with the same word.
    """

    def __init__(self, reason: str) -> None:
        super().__init__(reason)
        self.reason = reason


@dataclass
class Rgb:
    red: int = 0
    green: int = 0
    blue: int = 0

    def as_tuple(self) -> tuple:
        return (self.red, self.green, self.blue)


def _default_lights() -> List[Rgb]:
    return [Rgb() for _ in range(LIGHT_COUNT)]


@dataclass
class LightsCommand:
    sequence: int = 0
    lights: List[Rgb] = field(default_factory=_default_lights)


@dataclass
class LightsState:
    sequence: int = 0
    result: int = LightsResult.OK
    active_source: int = LightsSource.NONE
    lights: List[Rgb] = field(default_factory=_default_lights)


@dataclass
class EstopClearResetRequest:
    sequence: int = 0


@dataclass
class EstopClearResetState:
    sequence: int = 0
    result: int = EstopClearResetResult.OK
    needs_reset: int = 0
    estop_active: int = 0


@dataclass
class PinoutEvent:
    """A vehicle-originated pinout snapshot.

    Not a reply, so there is no header sequence to correlate with; the counter
    that matters is ``event_sequence``. ``uptime_ms`` is milliseconds since the
    vehicle booted and is not comparable to ROS time, and ``boot_id`` is opaque:
    compare it for equality to notice a restart, never order it.

    The ``_changed`` masks are sticky since the previous event, so a pulse that
    came and went between two publications is still visible. A bit outside its
    ``_valid`` mask is rejected, which is what keeps "not wired" distinct from
    "not asserted".
    """

    uptime_ms: int = 0
    event_sequence: int = 0
    boot_id: int = 0
    inputs_level: int = 0
    inputs_changed: int = 0
    inputs_valid: int = 0
    outputs_level: int = 0
    outputs_changed: int = 0
    outputs_valid: int = 0


def _check_pinout_masks(level: int, changed: int, valid: int) -> None:
    for value, what in ((level, "level"), (changed, "changed"), (valid, "valid")):
        if not isinstance(value, int) or isinstance(value, bool):
            raise WireError(f"pinout {what} mask must be an integer")
        if not 0 <= value <= _UINT64_MAX:
            raise WireError(f"pinout {what} mask out of range")
    if (level & ~valid) or (changed & ~valid):
        raise WireError("INVALID_PINOUT_MASK")


def _check_channel(value: int, what: str) -> int:
    if not isinstance(value, int) or isinstance(value, bool):
        raise WireError(f"{what} must be an integer")
    if not 0 <= value <= 255:
        raise WireError(f"{what} out of range 0..255")
    return value


def _encode_header(message_type: int, payload_size: int, sequence: int) -> bytes:
    if not 0 <= sequence <= 0xFFFFFFFF:
        raise WireError("sequence out of range")
    return _HEADER.pack(MAGIC, PROTOCOL_VERSION, int(message_type), payload_size, 0, sequence)


def _decode_header(data: bytes, expected_type: int, expected_payload: int) -> int:
    """Validate a header and return its sequence.

    Checks run in the same order as the C decoder, so an identical frame is
    rejected for an identical reason on either side.
    """
    if len(data) != HEADER_SIZE + expected_payload:
        raise WireError("INVALID_LENGTH")
    magic, version, message_type, payload_size, _reserved, sequence = _HEADER.unpack(
        data[:HEADER_SIZE]
    )
    if magic != MAGIC:
        raise WireError("INVALID_MAGIC")
    if version != PROTOCOL_VERSION:
        raise WireError("INVALID_VERSION")
    if message_type != expected_type:
        raise WireError("INVALID_MESSAGE_TYPE")
    if payload_size != expected_payload:
        raise WireError("INVALID_LENGTH")
    return sequence


def _encode_rgb(lights: List[Rgb]) -> bytes:
    if len(lights) != LIGHT_COUNT:
        raise WireError("INVALID_LIGHT_COUNT")
    out = bytearray()
    for index, colour in enumerate(lights):
        out.append(_check_channel(colour.red, f"lights[{index}].red"))
        out.append(_check_channel(colour.green, f"lights[{index}].green"))
        out.append(_check_channel(colour.blue, f"lights[{index}].blue"))
    return bytes(out)


def _decode_rgb(payload: bytes) -> List[Rgb]:
    return [
        Rgb(payload[i * 3], payload[i * 3 + 1], payload[i * 3 + 2])
        for i in range(LIGHT_COUNT)
    ]


def encode_lights_command(command: LightsCommand) -> bytes:
    body = bytes([LIGHT_COUNT, 0, 0, 0]) + _encode_rgb(command.lights)
    return _encode_header(MessageType.LIGHTS_COMMAND, LIGHTS_PAYLOAD_SIZE, command.sequence) + body


def decode_lights_command(data: bytes) -> LightsCommand:
    sequence = _decode_header(data, MessageType.LIGHTS_COMMAND, LIGHTS_PAYLOAD_SIZE)
    if data[HEADER_SIZE] != LIGHT_COUNT:
        raise WireError("INVALID_LIGHT_COUNT")
    return LightsCommand(sequence=sequence, lights=_decode_rgb(data[HEADER_SIZE + 4 :]))


def encode_lights_state(state: LightsState) -> bytes:
    if not 0 <= int(state.result) <= int(LightsResult.BUSY_SOURCE):
        raise WireError("INVALID_RESULT")
    if not 0 <= int(state.active_source) <= int(LightsSource.EDUCATION_JSON):
        raise WireError("INVALID_SOURCE")
    body = bytes([LIGHT_COUNT, int(state.result), int(state.active_source), 0])
    body += _encode_rgb(state.lights)
    return _encode_header(MessageType.LIGHTS_STATE, LIGHTS_PAYLOAD_SIZE, state.sequence) + body


def decode_lights_state(data: bytes) -> LightsState:
    sequence = _decode_header(data, MessageType.LIGHTS_STATE, LIGHTS_PAYLOAD_SIZE)
    if data[HEADER_SIZE] != LIGHT_COUNT:
        raise WireError("INVALID_LIGHT_COUNT")
    result = data[HEADER_SIZE + 1]
    source = data[HEADER_SIZE + 2]
    if result > int(LightsResult.BUSY_SOURCE):
        raise WireError("INVALID_RESULT")
    if source > int(LightsSource.EDUCATION_JSON):
        raise WireError("INVALID_SOURCE")
    return LightsState(
        sequence=sequence,
        result=result,
        active_source=source,
        lights=_decode_rgb(data[HEADER_SIZE + 4 :]),
    )


def encode_pinout_event(event: PinoutEvent) -> bytes:
    _check_pinout_masks(event.inputs_level, event.inputs_changed, event.inputs_valid)
    _check_pinout_masks(event.outputs_level, event.outputs_changed, event.outputs_valid)
    if not 0 <= event.uptime_ms <= _UINT64_MAX:
        raise WireError("uptime_ms out of range")
    if not 0 <= event.event_sequence <= 0xFFFFFFFF:
        raise WireError("event_sequence out of range")
    if not 0 <= event.boot_id <= 0xFFFFFFFF:
        raise WireError("boot_id out of range")
    # Header sequence is zero: an event answers no request.
    header = _encode_header(MessageType.PINOUT_EVENT, PINOUT_EVENT_PAYLOAD_SIZE, 0)
    return header + _PINOUT_BODY.pack(
        event.uptime_ms,
        event.event_sequence,
        event.boot_id,
        event.inputs_level,
        event.inputs_changed,
        event.inputs_valid,
        event.outputs_level,
        event.outputs_changed,
        event.outputs_valid,
    )


def decode_pinout_event(data: bytes) -> PinoutEvent:
    _decode_header(data, MessageType.PINOUT_EVENT, PINOUT_EVENT_PAYLOAD_SIZE)
    fields = _PINOUT_BODY.unpack(data[HEADER_SIZE:])
    event = PinoutEvent(*fields)
    _check_pinout_masks(event.inputs_level, event.inputs_changed, event.inputs_valid)
    _check_pinout_masks(event.outputs_level, event.outputs_changed, event.outputs_valid)
    return event


def encode_estop_clear_reset_request(request: EstopClearResetRequest) -> bytes:
    header = _encode_header(
        MessageType.ESTOP_CLEAR_RESET_REQUEST, ESTOP_CLEAR_RESET_PAYLOAD_SIZE, request.sequence
    )
    return header + struct.pack("<I", 0)


def decode_estop_clear_reset_request(data: bytes) -> EstopClearResetRequest:
    sequence = _decode_header(
        data, MessageType.ESTOP_CLEAR_RESET_REQUEST, ESTOP_CLEAR_RESET_PAYLOAD_SIZE
    )
    return EstopClearResetRequest(sequence=sequence)


def encode_estop_clear_reset_state(state: EstopClearResetState) -> bytes:
    if not 0 <= int(state.result) <= int(EstopClearResetResult.INTERNAL_ERROR):
        raise WireError("INVALID_ESTOP_CLEAR_RESET_RESULT")
    header = _encode_header(
        MessageType.ESTOP_CLEAR_RESET_RESULT, ESTOP_CLEAR_RESET_PAYLOAD_SIZE, state.sequence
    )
    return header + bytes(
        [int(state.result), int(state.needs_reset), int(state.estop_active), 0]
    )


def decode_estop_clear_reset_state(data: bytes) -> EstopClearResetState:
    sequence = _decode_header(
        data, MessageType.ESTOP_CLEAR_RESET_RESULT, ESTOP_CLEAR_RESET_PAYLOAD_SIZE
    )
    result = data[HEADER_SIZE]
    if result > int(EstopClearResetResult.INTERNAL_ERROR):
        raise WireError("INVALID_ESTOP_CLEAR_RESET_RESULT")
    return EstopClearResetState(
        sequence=sequence,
        result=result,
        needs_reset=data[HEADER_SIZE + 1],
        estop_active=data[HEADER_SIZE + 2],
    )
